hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java

```diff
@@ -63,72 +65,97 @@ public HoodieMergeHandle(HoodieWriteConfig config,
-    this.writeStatus = writeStatus;
+    // Load the new records in a map
     this.keyToNewRecords = new HashMap<>();
+    while (newRecordsItr.hasNext()) {
+      HoodieRecord<T> record = newRecordsItr.next();
+      // If the first record, we need to extract some info out
+      initInternal(fileId, record);
+      keyToNewRecords.put(record.getRecordKey(), record);
+      // update the new location of the record, so we know where to find it next
+      record.setNewLocation(new HoodieRecordLocation(commitTime, fileId));
+    }
+  }
 
-    try {
-      // Load the new records in a map
-      while (newRecordsItr.hasNext()) {
-        HoodieRecord<T> record = newRecordsItr.next();
-        // If the first record, we need to extract some info out
-        if (oldFilePath == null) {
-          String latestValidFilePath = fileSystemView
-              .getLatestDataFiles(record.getPartitionPath())
-              .filter(dataFile -> dataFile.getFileId().equals(fileId))
-              .findFirst()
-              .get().getFileName();
-          writeStatus.getStat().setPrevCommit(FSUtils.getCommitTime(latestValidFilePath));
-
-          HoodiePartitionMetadata partitionMetadata = new HoodiePartitionMetadata(fs,
-              commitTime,
-              new Path(config.getBasePath()),
-              new Path(config.getBasePath(), record.getPartitionPath()));
-          partitionMetadata.trySave(TaskContext.getPartitionId());
-
-          oldFilePath = new Path(
-              config.getBasePath() + "/" + record.getPartitionPath() + "/"
-                  + latestValidFilePath);
-          String relativePath = new Path(record.getPartitionPath() + "/" + FSUtils
-              .makeDataFileName(commitTime, TaskContext.getPartitionId(), fileId)).toString();
-          newFilePath = new Path(config.getBasePath(), relativePath);
-
-          // handle cases of partial failures, for update task
-          if (fs.exists(newFilePath)) {
-            fs.delete(newFilePath, false);
-          }
-
-          logger.info(String.format("Merging new data into oldPath %s, as newPath %s",
-              oldFilePath.toString(), newFilePath.toString()));
-          // file name is same for all records, in this bunch
-          writeStatus.setFileId(fileId);
-          writeStatus.setPartitionPath(record.getPartitionPath());
-          writeStatus.getStat().setFileId(fileId);
-          writeStatus.getStat().setPath(relativePath);
+  /**
+   * Initialize the incoming records map and extract the old file path.
+   */
+  private void init(String fileId, Map<String, HoodieRecord<T>> keyToNewRecords) {
+    this.keyToNewRecords = keyToNewRecords;
+    if (keyToNewRecords.size() > 0) {
+      initInternal(fileId, keyToNewRecords.values().stream().findFirst().get());
+    }
+  }
+
+  /**
+   * Initialize write status and extract old file path
+   * @param fileId
+   * @param record
+   */
+  private void initInternal(String fileId, HoodieRecord record) {
```

Comment:
why the refactor and two init methods? is n't all that was needed was adding a constructor that sets the map directly?

Reply:
So the problem was that the current init() method by default iterates over the records and populates the map. So either we can have a boolean passed to the method and write if-else to iterate over the values or not, that seemed dirty. So, I separated out the initializing of metadata fields in one init() method and the map vs iterator stuff in another. Just take a look at the current code, it'll make sense.

Reply:
rename : initStorageWriter() and call if after you have the merged map in both constructors.